base/gmp.jl

```diff
@@ -30,43 +30,85 @@ const GMP_BITS_PER_LIMB = gmp_bits_per_limb()
 # be used whenever mp_limb_t is in the signature of ccall'ed GMP functions.
 if GMP_BITS_PER_LIMB == 32
     typealias Limb UInt32
+    typealias SLimbMax Union{Int8, Int16, Int32}
+    typealias ULimbMax Union{UInt8, UInt16, UInt32}
 elseif GMP_BITS_PER_LIMB == 64
     typealias Limb UInt64
+    typealias SLimbMax Union{Int8, Int16, Int32, Int64}
+    typealias ULimbMax Union{UInt8, UInt16, UInt32, UInt64}
 else
     error("GMP: cannot determine the type mp_limb_t (__gmp_bits_per_limb == $GMP_BITS_PER_LIMB)")
 end
 
+immutable BigInt <: Integer
+    size::Int
+    d::Vector{Limb}
+end
+
+BigInt() = BigInt(0)
 
-type BigInt <: Integer
+type MPZ <: Integer
     alloc::Cint
     size::Cint
     d::Ptr{Limb}
-    function BigInt()
-        b = new(zero(Cint), zero(Cint), C_NULL)
-        ccall((:__gmpz_init,:libgmp), Void, (Ptr{BigInt},), &b)
-        finalizer(b, _gmp_clear_func)
-        return b
-    end
+
+    MPZ() = new(zero(Cint), zero(Cint), C_NULL)
+    # global objects can't be initialized via libgmp in a constructor, hence the init() function
+end
+
+function init(z::MPZ)
+    ccall((:__gmpz_init,:libgmp), Void, (Ptr{MPZ},), &z)
+    # finalizing is probably unnecessary if init'ed MPZ object are only const global
+    finalizer(z, cglobal((:__gmpz_clear, :libgmp)))
+end
+
+# to wrap an existing BigInt x as an MPZ, we call a global MPZ object _W on x: _W(x)
+# the syntax MPZ(x) would work, but would allocate a new MPZ on the heap each time
+function (z::MPZ)(x::BigInt)
```

Comment:
Ah yes, I had meant to remove the `Integer` inheritance from `MPZ`! more seriously, I chose this syntax for its lightness, `getindex` would do it too. I'm fine with using a normal function, but I believe this would uncessarily overload the many places where it is used.